composable-tasks/src/compose_tasks.rs

```diff
@@ -161,6 +166,102 @@ pub trait TaskCall:
         EgldOrEsdtTokenPayment::from(payment_out)
     }
 
+    fn smart_swap(
+        &self,
+        payment_for_current_task: EgldOrEsdtTokenPayment,
+        payments_to_return: &mut PaymentsVec<Self::Api>,
+        args: &mut ManagedVec<ManagedBuffer>,
+    ) -> EgldOrEsdtTokenPayment {
+        require!(
+            !payment_for_current_task.token_identifier.is_egld(),
+            "EGLD can't be swapped!"
+        );
+        let mut payment_in = payment_for_current_task.unwrap_esdt();
+
+        let args_cloned = args.clone();
+        let mut aggregated_payment_out =
+            self.get_empty_payment_out_from_smart_swap_args(args_cloned);
+
+        let mut args_iter = args.into_iter();
+
+        loop {
+            let routes_no = match args_iter.next() {
+                Some(count) => count.parse_as_u64().unwrap(),
+                None => break,
+            };
+
+            // take the input amount for the swap
+            let amount_in = match args_iter.next() {
+                Some(amount) => BigUint::from(amount),
+                None => break,
+            };
+
+            payment_in.amount = amount_in;
+
+            // take args for a router_swap
+            let router_args: ManagedVec<ManagedBuffer> = args_iter
+                .by_ref()
+                .take(ROUTER_SWAP_ARGS_LEN * routes_no as usize)
+                .collect();
+
+            let mut returned_payments_by_router =
+                self.multi_pair_swap(payment_in.clone(), router_args);
+            require!(
+                !returned_payments_by_router.is_empty(),
+                "Router swap returned 0 payments"
+            );
+
+            // aggregate all output_payments
+            let partial_payment_out_index = returned_payments_by_router.len() - 1;
+            let partial_payment_out = returned_payments_by_router.take(partial_payment_out_index);
+            require!(
+                partial_payment_out.token_identifier == aggregated_payment_out.token_identifier,
+                "Router returned wrong payment output for smart swaps"
+            );
+            aggregated_payment_out.amount += partial_payment_out.amount;
+
+            // concatenate all the other payments
+            payments_to_return.append_vec(returned_payments_by_router);
+        }
+
+        let smart_swap_fee_percentage_mapper = self.smart_swap_fee_percentage();
+        if !smart_swap_fee_percentage_mapper.is_empty() {
+            let fee_percentage = smart_swap_fee_percentage_mapper.get();
+            let fee_amount =
+                self.calculate_fee_amount(&aggregated_payment_out.amount, fee_percentage);
+            aggregated_payment_out.amount -= &fee_amount;
+        }
```

Comment:
```suggestion
        let fee_percentage = self.smart_swap_fee_percentage().get();
        if fee_percentage != 0 {
            let fee_amount =
                self.calculate_fee_amount(&aggregated_payment_out.amount, fee_percentage);
            aggregated_payment_out.amount -= &fee_amount;
        }
```

Checking if it's empty and then getting the value actually costs more. Just get the value and compare to 0.